File: web/src/lib/session-state.ts
/**
 * Session state contract for the muxterm "home" view.
 *
 * This file is the browser-side mirror of internal/sessiond/sessionstate.go.
 * It is committed to the base commit ahead of implementation so the backend and
 * frontend can be built independently without either guessing the other's
 * shape. If you change a field here, change it there in the same commit.
 *
 * This contract is HARNESS-AGNOSTIC. Nothing here names a specific
 * coding-agent CLI: an Amplifier lane, a Claude Code session, and a shell
 * script that shelled out to `muxterm session report` are all just rows. The
 * producer contract lives in docs/session-state-protocol.md.
 *
 * Vocabulary note: the six nouns in play are workspace, pane, terminal,
 * session, project, and artifact. A "task" is not an object -- it is the
 * session's first prompt. Do not introduce new nouns.
 */

/**
 * Session lifecycle states, adopted verbatim from Claude Code's agent view so
 * the vocabulary matches what users already know.
 */
export type SessionRunState = 'working' | 'blocked' | 'done' | 'failed' | 'stopped';

/** Reasons a session is blocked. Only meaningful when state === 'blocked'. */
export type WaitingFor =
  | 'permission prompt'
  | 'input needed'
  | 'sandbox request'
  | 'worker request'
  | 'dialog open';

/**
 * Session run modes. This distinction is load-bearing for the whole feature,
 * and it answers exactly one question:
 *
 *     Does going quiet mean "broke" or "resting"?
 *
 * - `interactive`: the session ends its turn and waits for a human. That is its
 *   CONTRACT, not a fault. A quiet interactive session must NEVER surface as an
 *   alarm.
 * - `autonomous`: the session runs a loop toward a stop condition of its own. A
 *   quiet autonomous session means the loop BROKE, and that IS the alarm.
 *
 * These names are deliberately harness-neutral. They were once spelled
 * `goal|plain`, after Amplifier's /goal command, which only names the
 * distinction correctly if you already know what /goal is. Claude Code has
 * background and foreground sessions; a job CLI has batch runs. The distinction
 * is universal; the Amplifier spelling was not.
 *
 * Getting this wrong makes every idle session look like an emergency, users
 * learn to ignore the indicator, and the home view becomes worthless.
 */
export type SessionMode = 'interactive' | 'autonomous';

/**
 * Coding-agent CLIs muxterm recognizes by name, mirroring the Harness*
 * constants in internal/sessiond/sessionstate.go (which are in turn the agent
 * catalog's names -- one vocabulary, not two).
 */
export const KNOWN_HARNESSES = ['amplifier', 'claude', 'codex', 'opencode'] as const;

export type KnownHarness = (typeof KNOWN_HARNESSES)[number];

/**
 * Which agent CLI is running a session.
 *
 * Deliberately OPEN, not a closed union: any producer may declare any harness
 * string (see docs/session-state-protocol.md). `KnownHarness | (string & {})`
 * keeps editor autocomplete for the four muxterm knows about while still
 * accepting a fifth nobody has written yet. A value not in KNOWN_HARNESSES gets
 * a neutral badge and is NEVER dropped -- refusing to show a session because
 * muxterm has not heard of its runner would make the fleet view lie about the
 * fleet.
 */
export type Harness = KnownHarness | (string & {});

/** Whether a declared harness is one muxterm has a name for. */
export function isKnownHarness(h: string | undefined): h is KnownHarness {
  return !!h && (KNOWN_HARNESSES as readonly string[]).includes(h);
}

/**
 * One row of the home view: everything known about a single agent session
 * running in a muxterm pane.
 *
 * Every field is DECLARED by that session's own producer. Nothing here is
 * inferred from PTY state -- the daemon's activity classifier cannot
 * distinguish "thinking" from "waiting for you", which is why this declared
 * channel exists.
 */
export interface SessionState {
  /** The producer's own session id. */
  sessionId: string;
  /** muxterm pane holding this session's terminal. */
  paneId: number;
  /** muxterm workspace containing that pane. */
  workspaceId: string;
  /** Which agent CLI is running this. Absent means nothing was declared. */
  harness?: Harness;
  /** Working directory. Absolute path. */
  project?: string;
  /** Short title, conventionally the session's first prompt, trimmed. */
  name: string;
  mode: SessionMode;
  state: SessionRunState;
  /** Present only when state === 'blocked'. */
  waitingFor?: WaitingFor;
  /** Short line describing current activity, refreshed cheaply. */
  doing?: string;
  /** The session's declared stop condition. Normally only when autonomous. */
  doneMeans?: string;
  /** Distinct artifact paths this session has read. */
  knows?: string[];
  /**
   * Associated PR number. Shown on the row; it does NOT change the row's
   * group. Having a PR says nothing about whether the session wants you,
   * which is the only question the groups answer.
   */
  pr?: number;
  /** Unix timestamp (seconds) of the last state change. */
  updatedAt: number;
}

/**
 * Home view groups, in display order.
 *
 * A lifecycle, not a taxonomy. Anything you launch enters `Running`, and
 * leaves it in exactly one of two directions: it wants you (`Needs input`) or
 * it is over (`Completed`). There is no third exit and no state that sits
 * outside the story, which is what makes the set readable at a glance.
 *
 * `Completed` merges done, failed and stopped on purpose. How a session ended
 * is a property of the row -- and the row's summary line says which -- not a
 * reason to sit somewhere else. Splitting them would mean scanning three
 * groups to answer "is anything still going?".
 */
export const HOME_GROUPS = ['Needs input', 'Running', 'Completed'] as const;

export type HomeGroup = (typeof HOME_GROUPS)[number];

/**
 * Place a session into its home-view group.
 *
 * Three groups, and the shape is a lifecycle rather than a taxonomy: anything
 * you launch enters `Running`, and leaves it in exactly one of two directions
 * -- it wants you (`Needs input`) or it is over (`Completed`). There is no
 * fourth place to land and no state that sits outside the story.
 *
 * `Ready for review` used to be a group of its own, promoted by an open PR.
 * It is gone: having a PR is a PROPERTY of a finished session, not a stage in
 * its life, and giving it a group meant one finished session outranked another
 * for a reason that had nothing to do with whether it needed you. The PR still
 * shows on the row, where a property belongs.
 *
 * The rule for the first group lives in needsInput() and is called from here
 * rather than repeated, so the group a row lands in and the number on the
 * Start card can never disagree about what "needs input" means.
 */
export function groupFor(s: SessionState): HomeGroup {
  if (needsInput(s)) return 'Needs input';
  if (s.state === 'working') return 'Running';
  return 'Completed';
}

/**
 * Whether a session belongs in "Needs input".
 *
 * Two ways in, and the second is the entire reason `mode` exists on the wire.
 *
 * 1. `blocked` -- it is sitting at a permission prompt or has asked something.
 *    Explicit, any mode.
 *
 * 2. An AUTONOMOUS session that is `stopped`. This is the load-bearing half of
 *    the mode contract: going quiet means "resting" for an interactive session
 *    and "the loop broke" for an autonomous one. An interactive session ending
 *    its turn rests at `stopped` and must NEVER surface here -- that is its
 *    contract, not a fault. An autonomous session reaching `stopped` did not
 *    reach its own stop condition: it hit a turn cap or was cancelled, and it
 *    is standing there with no verdict, waiting for somebody to decide whether
 *    to resume it, re-scope it, or let it go. Same state word, opposite
 *    meaning, and `mode` is the only thing that can tell them apart.
 *
 * `failed` is deliberately NOT here, even though an autonomous lane can only
 * reach it from its own goal evaluator. `failed` is a VERDICT -- the lane
 * finished and the answer was no. `stopped` is the ABSENCE of one. The row
 * that wants a human decision is the one nobody has decided yet; a verdict can
 * be read at leisure in "Completed", which is where the triage grouping
 * deliberately merges every ending.
 *
 * Known transient: a goal lane that is about to report `achieved` passes
 * through `autonomous` + `stopped` for the seconds its summary call takes, so
 * it can flash here before settling into "Completed". That is the direction
 * chosen on purpose -- the alternative is a lane that broke reading as a quiet
 * resting session, which HIDES a failure, and hiding is the one direction this
 * view must never fail in.
 */
export function needsInput(s: SessionState): boolean {
  if (s.state === 'blocked') return true;
  return s.mode === 'autonomous' && s.state === 'stopped';
}

/** Count of sessions needing input, for the sidebar Start card. */
export function needsInputCount(sessions: readonly SessionState[]): number {
  return sessions.reduce((n, s) => (needsInput(s) ? n + 1 : n), 0);
}

/**
 * Per-workspace needs-input counts, for the sidebar workspace card badges.
 * The Start card total is the sum of these, so the two can never disagree.
 */
export function needsInputByWorkspace(
  sessions: readonly SessionState[],
): Map<string, number> {
  const out = new Map<string, number>();
  for (const s of sessions) {
    if (!needsInput(s)) continue;
    out.set(s.workspaceId, (out.get(s.workspaceId) ?? 0) + 1);
  }
  return out;
}

/** Shorten an absolute project path for display, e.g. "~/workspace/muxterm". */
export function shortProject(project: string | undefined): string {
  if (!project) return '';
  const parts = project.replace(/\/+$/, '').split('/');
  return parts[parts.length - 1] ?? '';
}

/**
 * Development fixture. Lets the home view be built and viewed before the
 * daemon-side producer exists; delete the import once live data flows.
 *
 * Deliberately a MIXED FLEET: four harnesses, one of them a made-up name no
 * version of muxterm will ever recognize. The home view is a fleet view for
 * any coding-agent CLI, and a fixture that showed only Amplifier rows would let
 * a harness-specific assumption creep back in unnoticed. The `ci-runner` row is
 * the neutral-badge case, and it is here on purpose.
 */
export const FIXTURE_SESSIONS: SessionState[] = [
  {
    sessionId: 'fx-extract-muxops',
    paneId: 1,
    workspaceId: 'parity',
    harness: 'amplifier',
    project: '/home/ken/workspace/muxterm',
    name: 'extract-muxops',
    mode: 'autonomous',
    state: 'blocked',
    waitingFor: 'input needed',
    doing: 'loop stopped, awaiting a decision',
    doneMeans: 'CLI and MCP dispatch through one layer; go build clean',
    knows: [
      '/home/ken/workspace/muxterm/internal/mcp/run.go',
      '/home/ken/workspace/muxterm/cmd/muxterm/pane_cmd.go',
      '/home/ken/workspace/muxterm/AGENTS.md',
    ],
    updatedAt: 0,
  },
  {
    sessionId: 'fx-pr-412-rebase',
    paneId: 9,
    workspaceId: 'infra',
    harness: 'claude',
    project: '/home/ken/workspace/muxterm',
    name: 'pr-412-rebase',
    mode: 'autonomous',
    state: 'blocked',
    waitingFor: 'permission prompt',
    doing: 'force-push rewrites 4 commits',
    updatedAt: 0,
  },
  {
    sessionId: 'fx-pane-send-cli',
    paneId: 2,
    workspaceId: 'parity',
    harness: 'amplifier',
    project: '/home/ken/workspace/muxterm',
    name: 'pane-send-cli',
    mode: 'autonomous',
    state: 'working',
    doing: 'writing keys to bytes translation',
    updatedAt: 0,
  },
  {
    sessionId: 'fx-scrollback-parity',
    paneId: 3,
    workspaceId: 'parity',
    harness: 'codex',
    project: '/home/ken/workspace/muxterm',
    name: 'scrollback-parity',
    mode: 'autonomous',
    state: 'done',
    doing: 'MCP get_screen pages history now',
    pr: 51,
    updatedAt: 0,
  },
  {
    sessionId: 'fx-nightly-smoke',
    paneId: 4,
    workspaceId: 'infra',
    // Not a harness muxterm has ever heard of -- a shell script reporting via
    // `muxterm session report --harness ci-runner`. Neutral badge, real row.
    harness: 'ci-runner',
    project: '/home/ken/workspace/muxterm',
    name: 'nightly-smoke',
    mode: 'autonomous',
    state: 'working',
    doing: 'stage 3 of 6 — reconnect matrix',
    doneMeans: 'all six stages green',
    updatedAt: 0,
  },
  {
    sessionId: 'fx-design-notes',
    paneId: 7,
    workspaceId: 'cos',
    harness: 'claude',
    project: '/home/ken/workspace/muxterm',
    name: 'design-notes',
    mode: 'interactive',
    state: 'stopped',
    doing: 'turn ended, waiting for you -- normal, not an alarm',
    updatedAt: 0,
  },
];

File: internal/sessiond/sessionstate.go
package sessiond

import "encoding/json"

// Session state contract for the muxterm "home" view.
//
// This file is TYPES ONLY -- the wire contract shared by the daemon (which
// produces session state) and the browser (which renders it). It is committed
// to the base commit ahead of implementation so the backend and frontend can be
// built independently without either guessing the other's shape.
//
// The mirror of this file is web/src/lib/session-state.ts. If you change a
// field here, change it there in the same commit.
//
// This contract is HARNESS-AGNOSTIC. It began life shaped around Amplifier,
// but nothing below names a specific coding-agent CLI: any producer that can
// write a JSON file can appear in the home view. The on-disk producer contract
// is documented in docs/session-state-protocol.md, and the shipped producers
// are the Amplifier hook (modules/hooks-muxterm-session), the `muxterm session
// report` verb, and the opt-in Claude Code adapter.
//
// Vocabulary note: the six nouns in play are workspace, pane, terminal,
// session, project, and artifact. A "task" is not an object -- it is the
// session's first prompt. Do not introduce new nouns.

// Session lifecycle states. These are adopted verbatim from Claude Code's
// agent view so the vocabulary matches what users already know.
const (
	SessionStateWorking = "working"
	SessionStateBlocked = "blocked"
	SessionStateDone    = "done"
	SessionStateFailed  = "failed"
	SessionStateStopped = "stopped"
)

// Reasons a session is blocked. Only meaningful when State == blocked.
// Also adopted from Claude Code's waitingFor enum.
const (
	WaitingForPermission = "permission prompt"
	WaitingForInput      = "input needed"
	WaitingForSandbox    = "sandbox request"
	WaitingForWorker     = "worker request"
	WaitingForDialog     = "dialog open"
)

// Session run modes. This distinction is load-bearing for the whole feature,
// and it answers exactly one question:
//
//		Does going quiet mean "broke" or "resting"?
//
//	  - ModeInteractive: the session ends its turn and waits for a human. That is
//	    its CONTRACT, not a fault. A quiet interactive session must NEVER be
//	    surfaced as an alarm.
//	  - ModeAutonomous: the session runs a loop toward a stop condition of its
//	    own. A quiet autonomous session means the loop BROKE, and that IS the
//	    alarm worth showing.
//
// These names are deliberately harness-neutral. They were once spelled
// goal|plain, after Amplifier's /goal command, which only names the
// distinction correctly if you already know what /goal is. Claude Code has
// background and foreground sessions; a job CLI has batch runs and attended
// runs. The distinction is universal; the Amplifier spelling was not.
//
// Getting this wrong makes every idle session look like an emergency, users
// learn to ignore the indicator, and the home view becomes worthless.
const (
	ModeInteractive = "interactive"
	ModeAutonomous  = "autonomous"
)

// Harness identifiers: which coding-agent CLI is running this session.
//
// These are the SAME strings as the agent catalog's names (agent_catalog.go),
// which is deliberate -- muxterm has one vocabulary for "which agent CLI is
// this", not two that can drift apart. agent_catalog.go is defined in terms of
// these constants so the compiler enforces it.
//
// The field is OPEN, not an enum: any producer may declare any harness string
// (see docs/session-state-protocol.md). A value not listed here is rendered
// with a neutral badge, never dropped -- refusing to display a session because
// muxterm has not heard of its runner would make the fleet view a liar about
// the fleet.
const (
	HarnessAmplifier = "amplifier"
	HarnessClaude    = "claude"
	HarnessCodex     = "codex"
	HarnessOpenCode  = "opencode"
)

// ValidState reports whether s is one of the five lifecycle states.
func ValidState(s string) bool {
	switch s {
	case SessionStateWorking, SessionStateBlocked, SessionStateDone,
		SessionStateFailed, SessionStateStopped:
		return true
	}
	return false
}

// ValidMode reports whether m is one of the two run modes.
func ValidMode(m string) bool {
	return m == ModeInteractive || m == ModeAutonomous
}

// ValidWaitingFor reports whether w is one of the blocked reasons. The empty
// string is valid: it means "not blocked", which is most sessions most of the
// time.
func ValidWaitingFor(w string) bool {
	switch w {
	case "", WaitingForPermission, WaitingForInput, WaitingForSandbox,
		WaitingForWorker, WaitingForDialog:
		return true
	}
	return false
}

// TodoItem is one task exactly as its producer declared it.
type TodoItem struct {
	Text   string `json:"text"`
	Status string `json:"status"`
}

// TodoProgress is a session's progress through its own declared task list. It
// carries both summary numbers and the complete declared list.
// Items are optional for compatibility with snapshots written before the full
// list joined the protocol.
type TodoProgress struct {
	// Done is how many items are completed; Total is how many exist. Rendered
	// as "3/10". Total is never 0 in a published record -- a producer that
	// would emit 0/0 omits the whole object instead.
	Done  int `json:"done"`
	Total int `json:"total"`

	// Current is the in-progress item's text, in the producer's present-tense
	// form ("Cutting the release"). Empty when nothing is in progress, which
	// is a real state: a list that is all-pending or all-complete has an
	// honest fraction and no current item.
	Current string     `json:"current,omitempty"`
	Items   []TodoItem `json:"items,omitempty"`
}

// SessionState is one row of the home view: everything known about a single
// agent session. A terminal is an optional attachment.
//
// Every field is DECLARED by the session's own producer -- nothing here is
// inferred from PTY state, because the daemon's existing activity classifier
// cannot distinguish "thinking" from "waiting for you", which is precisely why
// this declared channel exists.
type SessionState struct {
	// Identity. SessionID is the producer's own session id. Zero PaneID and an
	// empty WorkspaceID mean the session has no muxterm terminal attachment;
	// MarshalJSON emits those sentinels as null on the wire.
	SessionID   string `json:"sessionId"`
	PaneID      int    `json:"paneId"`
	WorkspaceID string `json:"workspaceId"`

	// Harness names the coding-agent CLI running this session -- one of the
	// Harness* constants, or any other string a producer chooses to declare.
	// Empty means the producer declared nothing, which is allowed: the row
	// still renders, just without a badge.
	//
	// This is what turns the home view from an Amplifier feature into a fleet
	// view: an Amplifier lane and a Claude Code session sit in the same list,
	// each labelled with what is actually running it.
	Harness string `json:"harness,omitempty"`
	// ExecutionID and TurnID carry native causal identity into lifecycle
	// notices. They are observations from hook reports, never inferred from
	// terminal activity or prose.
	ExecutionID string `json:"executionId,omitempty"`
	TurnID      string `json:"turnId,omitempty"`

	// Project is the session's working directory. Absolute path; the browser
	// shortens it for display.
	Project string `json:"project,omitempty"`

	// Name is a short human-readable title, conventionally derived from the
	// session's first prompt. This is the closest thing to a task title that
	// exists, and it costs nothing because a human typed it.
	Name string `json:"name"`

	// Label is a 1-3 word name for the work, short enough to fit a pane tab
	// whole -- "auth redirect", not the whole first line Name carries. The
	// producer derives it once, from its first prompt, and then never changes
	// it: a tab whose name moves is a tab you cannot find twice.
	//
	// Optional, and its absence is meaningful. Empty means this producer
	// offered nothing better than the label the daemon already derived from
	// the pane's launch argv at spawn (autolabel.go), so consumers keep what
	// they have rather than clearing it.
	Label string `json:"label,omitempty"`

	// Mode is ModeInteractive or ModeAutonomous. See the mode constants above.
	Mode string `json:"mode"`

	// State is one of the SessionState* constants.
	State string `json:"state"`

	// WaitingFor is one of the WaitingFor* constants. Empty unless blocked.
	WaitingFor string `json:"waitingFor,omitempty"`

	// Doing is a short human-readable line describing current activity, e.g.
	// "editing cmd/muxterm/pane_cmd.go". Refreshed cheaply from recent events.
	Doing string `json:"doing,omitempty"`

	// Summary is the lane's own raw final assistant message exactly as supplied
	// by its turn-end hook. It is separate from Doing so a detailed result can survive
	// without turning the fleet's one-line activity label into a transcript.
	Summary string `json:"summary,omitempty"`

	// DoneMeans is the session's own declared definition of finished -- the
	// stop condition an autonomous loop is running toward. Normally present
	// only when Mode == ModeAutonomous.
	DoneMeans string `json:"doneMeans,omitempty"`

	// GoalID and Origin are the two fields on this row that a producer does
	// NOT declare. They are stamped by the daemon during the pane join
	// (sessionstore.go stampPane) out of muxterm's own launch record, and
	// anything a producer writes into them is discarded -- a session cannot be
	// authoritative about the command line it was started with.
	//
	// GoalID identifies the stop condition this lane was LAUNCHED with, as a
	// digest of the condition text (lane_provenance.go GoalID). It answers two
	// questions DoneMeans cannot. First, correlation: two lanes running the
	// same condition in two workspaces share an id, so a batch is recognisable
	// as a batch. Second, survival: DoneMeans is declared by the session and
	// goes away the moment a human takes over a finished goal lane and it
	// stops calling itself autonomous -- GoalID is read from the pane's argv
	// and does not, so "which goal was this lane for?" stays answerable after
	// the conversation has moved on.
	//
	// It names the CONDITION, not the run: two runs of one condition share an
	// id. Empty for any lane not launched as a goal lane.
	GoalID string `json:"goalId,omitempty"`

	// Origin is which door this lane came through: browser, agent, cli, or
	// trigger:<trigger id> (the LaneOrigin* constants). Before this existed, a
	// lane a human spawned and a lane an automation fired at 3am were the same
	// row, and "which of these did I ask for?" could not be answered from the
	// fleet at all.
	//
	// Empty is a real value: a pane restored after a daemon restart came
	// through no door in this process's lifetime, and guessing would put a
	// wrong answer on the one question the field exists for.
	Origin string `json:"origin,omitempty"`

	// Todo is structured progress through the session's own task list, when it
	// keeps one. Nil means it does not, which is NOT the same as "no progress":
	// consumers must fall back to Doing rather than render an empty 0/0, or a
	// session that simply does not track todos reads as a stalled one.
	//
	// This is the only field on the row that is not a guess. Doing is
	// re-templated on every tool call and describes the last thing touched;
	// Todo changes only when the session itself revises its plan, which makes
	// it both cheaper to carry and far more stable to look at.
	Todo *TodoProgress `json:"todo,omitempty"`

	// Knows lists distinct artifact paths this session has read. A session
	// that has read very little and then failed was starved, not merely
	// unlucky, and that distinction is invisible without this.
	Knows []string `json:"knows,omitempty"`

	// PR is a pull request number associated with this session, if any.
	//
	// It is a PROPERTY of the row, not a group. "Ready for review" used to be
	// a group of its own promoted by this field; it was removed because having
	// a PR says nothing about whether the session wants you, which is the only
	// question the groups answer.
	PR int `json:"pr,omitempty"`

	// UpdatedAt is a Unix timestamp (seconds) of the last state change.
	UpdatedAt int64 `json:"updatedAt"`

	// Reporting is launch/report delivery health, not harness run state.
	Reporting         string `json:"reporting,omitempty"`
	ReportingCoverage string `json:"reportingCoverage,omitempty"`
	ReportingError    string `json:"reportingError,omitempty"`
	LastReportAt      int64  `json:"lastReportAt,omitempty"`
}

// MarshalJSON keeps the established internal scalar representation while
// making terminal attachment explicitly optional on the versioned wire. This
// is a compatibility bridge for producers and daemon code that still use zero
// values internally; consumers never receive a synthetic pane 0 or workspace.
func (s SessionState) MarshalJSON() ([]byte, error) {
	type wireSessionState SessionState
	var paneID *int
	var workspaceID *string
	if s.PaneID != 0 {
		paneID = &s.PaneID
	}
	if s.WorkspaceID != "" {
		workspaceID = &s.WorkspaceID
	}
	return json.Marshal(struct {
		wireSessionState
		PaneID      *int    `json:"paneId"`
		WorkspaceID *string `json:"workspaceId"`
	}{wireSessionState: wireSessionState(s), PaneID: paneID, WorkspaceID: workspaceID})
}

// NeedsInput reports whether this session belongs in the home view's
// "Needs input" group.
//
// Two ways in, and the second is the entire reason Mode exists on the wire:
//
//  1. Blocked -- sitting at a permission prompt, or having asked something.
//  2. An AUTONOMOUS session that is Stopped. Going quiet means "resting" for
//     an interactive session and "the loop broke" for an autonomous one. An
//     interactive session ending its turn rests at Stopped and must NEVER
//     surface here; an autonomous one reaching Stopped did not reach its own
//     stop condition and is waiting for somebody to decide what happens next.
//
// Failed is deliberately absent: it is a VERDICT, and a verdict can be read at
// leisure in "Completed". Stopped is the ABSENCE of one, which is what wants a
// human.
//
// This is the Go mirror of needsInput() in web/src/lib/session-state.ts, which
// is where the rule is actually CONSUMED -- grouping happens in the browser.
// Nothing in this package calls this today; it exists so the pinned contract
// states the rule in both languages. If you change one, change the other.
func (s SessionState) NeedsInput() bool {
	if s.State == SessionStateBlocked {
		return true
	}
	return s.Mode == ModeAutonomous && s.State == SessionStateStopped
}
